engine/calculus/LinearSpline1D.cs

```diff
@@ -0,0 +1,242 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Numerics;
+using System.Linq;
+using System;
+
+namespace FreedomOfFormFoundation.AnatomyEngine.Calculus
+{
+	/// <summary>
+	///     Class <c>LinearSpline1D</c> describes a one-dimensional linear spline, which is a piecewise function.
+	///		Each piece is defined by a linear function, \f$q(x) = a_0 + a_1 x, for which the
+	///		parameters are defined such that the piecewise function is continuous. A spline is defined by a series of
+	///		points that the function must intersect, and the program will automatically generate a curve that passes
+	///		through these points. A linear spline is discontinuous on its first derivative.
+	/// </summary>
+	public class LinearSpline1D : RaytraceableFunction1D
+	{
+		private float[] _parameters;
+		public SortedPointsList<float> Points { get; }
+		
+		/// <summary>
+		///     Construct a linear spline using a set of input points.
+		/// 	<example>For example:
+		/// 	<code>
+		/// 		SortedList<float, float> splinePoints = new SortedList<float, float>();
+		/// 		splinePoints.Add(0.0f, 1.1f);
+		/// 		splinePoints.Add(0.3f, 0.4f);
+		/// 		splinePoints.Add(1.0f, 2.0f);
+		/// 		LinearSpline1D spline = new LinearSpline1D(splinePoints);
+		/// 	</code>
+		/// 	creates a linear spline that passes through three points: (0.0, 1.1), (0.3, 0.4) and (1.0, 2.0).
+		/// 	</example>
+		/// </summary>
+		/// <param name="points">A list of points that is sorted by the x-coordinate. This collection is copied.</param>
+		/// <exception cref="ArgumentException">
+		/// 	A linear spline must have at least two points to be properly defined. If <c>points</c> contains less than
+		/// 	two points, the spline is undefined, so an <c>ArgumentException</c> is thrown.
+		/// </exception>
+		public LinearSpline1D(SortedList<float, float> points)
+		{
+			if (points.Count < 2)
+			{
+				if (points.Count == 1)
+				{
+					throw new ArgumentException("List contains only a single point. A spline must have at least two points.", "points");
+				}
+				else
+				{
+					throw new ArgumentException("List is empty. A spline must have at least two points.", "points");
+				}
+			}
+			
+			Points = new SortedPointsList<float>(points);
+
+			// Calculate the coefficients for each segment of the spline:
+			_parameters = new float[points.Count];
+
+			// Find the first segment parameter, which will be a straight line:
+			{
+				float dx = Points.Key[1] - Points.Key[0];
+				float dy = Points.Value[1] - Points.Value[0];
+
+				_parameters[0] = dy / dx;
+			}
+
+			// Recursively find the other _parameters:
+			for (int i = 1; i < points.Count; i++)
+			{
+				float dx = Points.Key[i] - Points.Key[i - 1];
+				float dy = Points.Value[i] - Points.Value[i - 1];
+
+				_parameters[i] = dy / dx;
+			}
```

Comment:
This seems to be able to be condensed into a single loop. Let me know if that's indeed the case!

```suggestion
			// Recursively find the _parameters:
			for (int i = 0; i < points.Count; i++)
			{
				float dx = Points.Key[i] - Points.Key[i - 1];
				float dy = Points.Value[i] - Points.Value[i - 1];

				_parameters[i] = dy / dx;
			}
```

You simply start at `i=0` instead of `i=1`, so that you get the first iteration for free. If I recall correctly, the reason that this was different in the other splines was because the first segment was a special case and needed to be a line segment instead of a higher-order curve, but since they're all line segments in this case, that doesn't matter anymore of course.

Reply:
nice, thanks!